--- nfdump-1.6.13/bin/nffile.c	2014-03-12 10:45:49.000000000 +0100
+++ nfdump/bin/nffile.c	2015-03-08 08:23:26.446789688 +0100
@@ -77,11 +77,11 @@
     lzo_align_t __LZO_MMODEL var [ ((size) + (sizeof(lzo_align_t) - 1)) / sizeof(lzo_align_t) ]
 
 static HEAP_ALLOC(wrkmem,LZO1X_1_MEM_COMPRESS);
-static void *lzo_buff;
-static int lzo_initialized = 0;
+static __thread void *lzo_buff;
+static __thread int lzo_initialized = 0;
 
 #define ERR_SIZE 256
-static char	error_string[ERR_SIZE];
+static __thread char	error_string[ERR_SIZE];
 
 static int LZO_initialize(void);
 
