--- nfdump-1.6.13/bin/nffile.c	2014-03-12 10:45:49.000000000 +0100
+++ nfdump/bin/nffile.c	2015-05-07 21:20:48.707836929 +0200
@@ -64,6 +64,13 @@
 
 /* global vars */
 
+#ifdef LNF_THREADS
+static __thread int open_files = 0;
+#else
+static int open_files = 0;
+#endif
+
+
 // required for idet filter in nftree.c
 char 	*CurrentIdent;
 
@@ -238,6 +247,8 @@
 		}
     }
 
+	open_files++;
+
 	return nffile;
 
 } // End of OpenFile
@@ -251,6 +262,15 @@
 	if ( nffile->fd )
 		close(nffile->fd);
 
+	open_files--;
+
+	if (open_files == 0 && lzo_initialized && lzo_buff) {
+		free(lzo_buff);
+		lzo_buff = NULL;
+		lzo_initialized = 0;
+	}
+
+
 } // End of CloseFile
 
 int ChangeIdent(char *filename, char *Ident) {
@@ -401,6 +421,8 @@
 	nffile->block_header->flags		 = 0;
 
 	nffile->buff_ptr = (void *)((pointer_addr_t)nffile->block_header + sizeof(data_block_header_t));
+
+	open_files++;
 	
 	return nffile;
 
